### DISCUSSION TITLE
esp32-cam module - error 0x105

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#18492] [Support request]
Hi All,
I bought a few cheap ESP32cam modules (logo is 'DM') - some time ago - mainly because they had an extrenal wifi/BLE antenna connection for BLE use.
They are probably [this](https://www.diymore.cc/products/esp32-cam-wifi-wireless-module-esp32-serial-to-wifi-esp32-cam-spi-flash-bluetooth-development-board-with-ov2640-camera-module#shopify-product-reviews) module.

Now I'd like to try the camera (I'm modifying a pet feeder to use ESP32 Tas, and fancied watching my fish feed....).

But...  I've tried flashing the current release () and on two of the modules... and get:

```
15:47:07.666 CAM: Stream init
15:47:07.669 CAM: Template pin config
15:47:07.669 CAM: XCLK on GPIO 0 using ledc channel 1
15:47:07.670 CAM: PSRAM found
15:47:07.706 CAM: Init failed with error 0x105
```
I've tried the 'AI Thinker' profile, and 'esp32 cam' profile - both the same.
The boards have only 5v connected - to a capable USB supply.

Is it possible that running them for some time with a generic profile could have damaged the camera module(s)?
Or is it possible these modules use different pins for the camera? (please suggest changes?)
Googling tells me that this is a camera connection issue.  I re-seated one, before trying the next module.

best regards,
Simon

Full log:
```
00:00:00.003 HDW: ESP32-D0WDQ6 (PSRAM)
00:00:00.027 UFS: FlashFS mounted with 36 kB free
00:00:00.035 CFG: Loaded from File, Count 103337
00:00:00.043 QPC: Count 1
00:00:00.047 SPI: Hardware using GPIO14(CLK), GPIO15(MOSI) and GPIO02(MISO)
00:00:00.058 BRY: GC from 3585 to 2779 bytes, objects freed 7/18 (in 0 ms) - slots from 34/61 to 18/61
00:00:00.098 CFG: No '*.autoconf' file found
00:00:00.102 BRY: GC from 3990 to 3395 bytes, objects freed 3/32 (in 1 ms) - slots from 46/76 to 31/76
00:00:00.102 BRY: Berry initialized, RAM used=3395 bytes
00:00:00.110 BRY: No 'preinit.be'0
00:00:00.721 SRC: Restart
00:00:00.726 Project tasmota - TasmotaStairs Version 12.5.0(webcam)-2_0_7(2023-04-17T08:06:59)
00:00:00.758 BRY: No 'autoexec.be'
00:00:00.029 WIF: Checking connection...
00:00:00.029 WIF: Attempting connection...
00:00:01.435 WIF: Connecting to AP1 MYSSID Channel 11 BSSId AC:B6:87:C7:0E:B8 in mode 11n as stairs...
00:00:02.206 WIF: Checking connection...
00:00:02.206 WIF: Attempting connection...
00:00:02.276 WIF: IPv4 192.168.1.190, mask 255.255.255.0, gateway 192.168.1.254
00:00:03.219 WIF: Checking connection...
00:00:03.219 WIF: Connected
00:00:03.526 HTP: Web server active on stairs with IP address 192.168.1.190
00:00:03.919 NTP: Sync time...
00:00:03.922 WIF: DNS failed for pool.ntp.org after 2 ms
00:00:04.287 WIF: IPv6 Local fe80::7e9e:bdff:fe37:f610
00:00:04.470 MQT: Attempting connection...
00:00:04.472 WIF: DNS failed for raspberrypi.lan after 1 ms
00:00:04.473 MQT: Connect failed to raspberrypi.lan:1883, rc -5. Retry in 10 sec
00:00:04.919 NTP: Sync time...
00:00:04.921 WIF: DNS failed for nl.pool.ntp.org after 2 ms
00:00:05.919 NTP: Sync time...
00:00:05.921 WIF: DNS failed for 0.nl.pool.ntp.org after 2 ms
00:00:06.925 QPC: Reset
00:00:06.926 NTP: Sync time...
00:00:06.928 WIF: DNS resolved '192.168.1.254' (192.168.1.254) in 0 ms
00:00:07.934 NTP: No reply from 192.168.1.254
00:00:07.945 NTP: Sync time...
00:00:07.948 WIF: DNS failed for 0.pool.ntp.org after 1 ms
00:00:08.291 WIF: IPv6 Global 2a00:23c6:663c:e601:7e9e:bdff:fe37:f610
00:00:08.941 APP: Boot Count 1971
00:00:08.941 NTP: Sync time...
00:00:08.943 WIF: DNS resolved 'pool.ntp.org' (85.199.214.102) in 1 ms
00:00:08.961 RTC: UTC 2023-04-23T14:47:03, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
15:47:03.000 RTC: Synced by NTP
15:47:03.223 CFG: Saved, Count 103338, Bytes 4096
15:47:07.596 HTP: Main Menu
15:47:07.666 CAM: Stream init
15:47:07.669 CAM: Template pin config
15:47:07.669 CAM: XCLK on GPIO 0 using ledc channel 1
15:47:07.670 CAM: PSRAM found
15:47:07.706 CAM: Init failed with error 0x105
15:47:10.701 MQT: Attempting connection...
15:47:10.704 WIF: DNS failed for raspberrypi.lan after 1 ms
15:47:10.705 MQT: Connect failed to raspberrypi.lan:1883, rc -5. Retry in 20 sec
15:47:18.454 WIF: Checking connection...
15:47:23.555 HTP: Consoles
15:47:25.386 HTP: Console
15:47:31.702 MQT: Attempting connection...
15:47:31.705 WIF: DNS failed for raspberrypi.lan after 1 ms
15:47:31.706 MQT: Connect failed to raspberrypi.lan:1883, rc -5. Retry in 30 sec
```

**Comment by Jason2866** [comment]
The AI Thinker clones and derivates are problematic devices. If you do some Internet research you will find many hits. Do NOT use the adapter board sold for. The power source needs to be rock solid and needs to deliver clean 5V. Most 5V USB Phone Chargers will not work. Remove after flashing the wires from all GPIOs 
GPIO 0 is used a clock input. Clock is some MHZ. So every piece of an wire there will kill the signal.
The error `Init failed with error 0x105 ` is the error message that happens when one of above is not fine.

**Comment by Jason2866** [comment]
I got a board with this bug 😜 https://community.home-assistant.io/t/esphome-esp32-wrover-e-camera-failed-solved/496863/4
Never had any problem using ESP32-S3 Cam boards...

**Reply by btsimonh** [reply]
Thanks Jason, I'll see if I can measure the voltages directly at the camera interface tonight, and try a different USB power supply, and report.  It feels like a manufacturing defect as two boards of the same batch behave exactly the same....

**Comment by btsimonh** [comment]
I checked voltages - they seem to be controlled via GPIO32 as expected, and seem to be the right voltages on the right pins (although my camera has pin numbers on it which seem reversed compared to the circuit diagram - the heavyier wires lead to the right places with the camera facing forward.
I note a new discussion with the same issue...  so maybe i'll try an older tas firmware....
I have nothing except power connected to be board, and have tried a different (5v@2A) supply.

**Comment by btsimonh** [comment]
ok, works on 10.0.0 (crashed after initial reboot after web 'upgrade', but worked after power cycle).
I will work forward through versions to see where it breaks....

**Comment by btsimonh** [comment]
I think I found why, and maybe solved it.
The esp driver toggles pwdn for 10ms to 'reset' the camera.
However, if the camera was not powered at boot, 10ms is not enough time for the voltages to stabilise and the camera to start running.
I added some code in the new driver to allow us to power down the camera, and then added a deliberate power up in WcStartup, with a 200ms delay, and now I can recover from a 0x105 (never could before without a hard power off).

More testing needed, but maybe we could go further and tell the esp driver that it does not have a pwdn pin, so we have complete control.

**Reply by mrneutron42** [reply]
@btsimonh did you find a conclusive fix for the InitErr 0x105 errors?  
Did a longer delay fix the problem?
If so, is this going to get submitted as a pull request into the main branch of Tasmota code?

I've got some the the ESP32-CAM boards with the DM on the metal cover.  After working fine for months, one of them started giving the InitErr 0x105 after a soft reboot this week.  After that, when I issue the 'wcinit' command, I get this: 
```
09:09:53.637 CMD: wcinit
09:09:53.677 CAM: InitErr 0x105 try 1
09:09:53.714 CAM: InitErr 0x105 try 2
09:09:53.751 CAM: InitErr 0x105 try 3
09:09:53.752 CAM: InitErr 0x105
09:09:53.789 CAM: InitErr 0x105 try 1
09:09:53.826 CAM: InitErr 0x105 try 2
09:09:53.863 CAM: InitErr 0x105 try 3
09:09:53.864 CAM: InitErr 0x105
```
I have the camera in a remote location and a 'restart 1' command is my only option of restarting it from a distance.  
Doing the 'restart 1' soft restart has not been enough to recover the functionality.
In the past, I think my workaround for this was to physically pull the power plug to the camera and do a hard reboot, but the problem occurred again months later.

**Reply by btsimonh** [reply]
@mrneutron42 - did you try the 'new' driver?

#define USE_WEBCAM_V2

Depending upon your use case, it should be more performant too.

**Reply by btsimonh** [reply]
I think my final fix was toggling the power on the cam module, around line 1020 of
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino
Relies on having a power GPO on the cam board.

**Reply by Aint-Nobody** [reply]
I try it with a custom compilation 14.5. with define use... V2
The compilation runs successful, 
but the fail error 0X105 is the same 😭🤔
Is there any new set option to increase the time of this critical timer?

**Reply by Jason2866** [reply]
105 is an error message which occurs for many issues. Software can't heal borderline hardware.
Dump this crap and save your time. If you want a worry free running cam, buy a S3 based one.

**Reply by Aint-Nobody** [reply]
@Jason2866 thanks for your answer. 
It clear my mind from confusion. 
Sd card also doesn't work on this board.
It seems ready for trashbin. 😄

**Reply by btsimonh** [reply]
Yep, my 'fix' for 0x105 was based on some boards having different uses of control lines for reset and power.  So specifically it helped with certain boards with odd wiring, but not with boards which just have bad wiring.

**Comment by mrneutron42** [comment]
Do you have to make a custom compile to use the "define USE_WEBCAM_V2"?
I've been using the stock compiles of the tasmota32-webcam.bin file from https://ota.tasmota.com/tasmota32/release/

I've got the common ESP32-CAM modules like this:
https://www.amazon.com/gp/product/B096WBMZ43/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&th=1
I think this is the schematic for the ESP32 camera board:
https://github.com/SeeedDocument/forum_doc/raw/master/reg/ESP32_CAM_V1.6.pdf
I see a FET connected to GPIO32 which switches the power to the CSI 2.8V and CSI 1.2V power rails.
So I think my boards would respond to the power switching GPIO.

**Reply by sfromis** [reply]
Yes, a custom build is needed to include a custom setting.

And with one of those crappy ESP32-CAM boards (cannot recommend) it still fails, but messages are different.
```
20:29:53.628 CMD: Grp 0, Cmd 'WCINIT', Idx 1, Len 0, Pld -99, Data ''
20:29:53.667 CAM: pwdn pin 32
20:29:53.869 CAM: InitErr 0x103 try 1
20:29:54.019 CAM: InitErr 0xffffffff try 2
20:29:54.168 CAM: InitErr 0xffffffff try 3
20:29:54.169 CAM: InitErr 0xffffffff
20:29:54.174 MQT: xxxxxxxx/stat/WCINIT = {"WCInit":"Done"}
```

A long time ago, I observed that `InitErr 0x105` was consistent with hardware failure, got the same symptom on a board where I later observed that the small ribbon cable to the camera was torn.

**Reply by btsimonh** [reply]
It's true that some of these boards are 'just a bit s**t'.  Speaking only for the boards I had at the time, I got no more failures after implementing the longer power off.  Other things: Whatever you do, don't mix high bandwidth wifi (like cam traffic) with BLE - things just don't go well at all.  If you suspect cam/wifi interaction (i.e. your wifi seems slow on one board compared to another), try 'de-tuning' the base cam frequency.

**Comment by mrneutron42** [comment]
[btsimonh](https://github.com/btsimonh), I'd love to test the reset delay in the custom firmware compile.
Can you share the file or a link to it?

I just made a trip to my remote camera site and simply did a 'hard reset' (pulled the power for 10 seconds) and the camera was recovered - no more 'InitErr 0x105' errors upon repower!  
To recap, doing a remote "soft reset" (issuing the 'restart 1' command) did not work.
Conclusion:  this is not a physical connection problem with my ESP32-CAM camera.

I suspect that the longer (200mS) power down in the modified code would help my recover my cameras.

**Comment by mrneutron42** [comment]
Can someone advise me about doing a Tasmota compile to use the V2 webcam driver?
Can I use TasmoCompiler?
reference: https://github.com/benzino77/tasmocompiler
Which checkboxes need to be checked (besides ESP32 webcam) and which custom parameters need to be entered into the box?
#define USE_WEBCAM_V2 ?

**Reply by sfromis** [reply]
Try with these in custom parameters:
```
#define USE_WEBCAM_V2
#define USE_BERRY_CAM
#define USE_BERRY_IMAGE
```
The two last may not be needed, are for extra features related to webcam.

I've not tested this using TasmoCompiler, but it seemed to work when building by other means.

**Reply by btsimonh** [reply]
not sure if the 2nd two will play nice with a (any) webcam driver :(.  It was tough to make the cam work in a separate thread...

**Reply by sfromis** [reply]
Having all 3 added to a webcam build, I'm still seeing at least basic webcam operation work. Not having done much further testing, I can't say if it plays nice.

**Reply by mrneutron42** [reply]
I'll start with only adding #define USE_WEBCAM_V2
The restart delay is what I'm really interested in trying.

**Comment by btsimonh** [comment]
hi @mrneutron42
I should really have been on this for you as I personally spend many weeks working on the V2 version.
But I've got a lot going on in my personal life, which gets in the way of TAS every time :(.
I think (not checked) that just adding the define you mention will work.
If I get a chance, I'll try to build one for you, but won't be tonight.
br,
Simon

**Reply by mrneutron42** [reply]
Thanks for the input @btsimonh and @sfromis.
I will make a custom compile with TasmoCompiler tonight.

**Comment by mrneutron42** [comment]
I used TasmoCompiler with the #define USE_WEBCAM_V2 parameter and made a 14.4.1 tasmota32-webcam firmware with V2 driver.
It works fine with the "DM" ESP32-CAM modules I have on my desk.
I just loaded it into one of my remote cameras and it is working fine.
Further testing showed there were no more 'InitErr 0x105' errors upon soft restart!
UPDATE:
I have also compiled a custom 14.5.0 tasmota32-webcam with the V2 driver and it also solves the 'InitErr 0x105' error problem.

Is this V2 driver eventually going to replace the V1 driver, and thus we won't have to make custom compiles to get the fixes and improvements of the V2 driver?

**Reply by wprinz** [reply]
This is just a confirmation - I just came across this discussion after trying to update an old Esp32Cam from v11.x to the new 14.5. x version. This failed with the InitErr problems that are reported above. After compiling tasmota32-webcam locally with VS Code and PlatformIO and **`#define USE_WEBCAM_V2`** in `user_config_override.h` the device works perfect. All the new commands listed here https://tasmota.github.io/docs/Commands/#camera can now be issued.
